src/pyclass.rs

```diff
@@ -0,0 +1,533 @@
+//! An experiment module which has all codes related only to #[pyclass]
+use crate::class::methods::{PyMethodDefType, PyMethodsProtocol};
+use crate::conversion::{AsPyPointer, FromPyPointer, ToPyObject};
+use crate::exceptions::RuntimeError;
+use crate::pyclass_slots::{PyClassDict, PyClassWeakRef};
+use crate::type_object::{type_flags, PyConcreteObject, PyTypeObject};
+use crate::types::PyAny;
+use crate::{class, ffi, gil, PyErr, PyObject, PyResult, PyTypeInfo, Python};
+use std::ffi::CString;
+use std::mem::ManuallyDrop;
+use std::os::raw::c_void;
+use std::ptr::{self, NonNull};
+
+/// A trait that enables custome alloc/dealloc implementations for pyclasses.
+pub trait PyClassAlloc: PyTypeInfo + Sized {
+    unsafe fn alloc(_py: Python) -> *mut Self::ConcreteLayout {
+        let tp_ptr = Self::type_object();
+        let alloc = (*tp_ptr).tp_alloc.unwrap_or(ffi::PyType_GenericAlloc);
+        alloc(tp_ptr, 0) as _
+    }
+
+    unsafe fn dealloc(py: Python, self_: *mut Self::ConcreteLayout) {
+        (*self_).py_drop(py);
+        let obj = self_ as _;
+        if ffi::PyObject_CallFinalizerFromDealloc(obj) < 0 {
+            return;
+        }
+
+        match Self::type_object().tp_free {
+            Some(free) => free(obj as *mut c_void),
+            None => tp_free_fallback(obj),
+        }
+    }
+}
+
+#[doc(hidden)]
+pub unsafe fn tp_free_fallback(obj: *mut ffi::PyObject) {
+    let ty = ffi::Py_TYPE(obj);
+    if ffi::PyType_IS_GC(ty) != 0 {
+        ffi::PyObject_GC_Del(obj as *mut c_void);
+    } else {
+        ffi::PyObject_Free(obj as *mut c_void);
+    }
+
+    // For heap types, PyType_GenericAlloc calls INCREF on the type objects,
+    // so we need to call DECREF here:
+    if ffi::PyType_HasFeature(ty, ffi::Py_TPFLAGS_HEAPTYPE) != 0 {
+        ffi::Py_DECREF(ty as *mut ffi::PyObject);
+    }
+}
+
+pub trait PyClass: PyTypeInfo + Sized + PyClassAlloc + PyMethodsProtocol {
+    type Dict: PyClassDict;
+    type WeakRef: PyClassWeakRef;
+}
+
+unsafe impl<T> PyTypeObject for T
+where
+    T: PyClass,
+{
+    fn init_type() -> NonNull<ffi::PyTypeObject> {
+        let type_object = unsafe { <Self as PyTypeInfo>::type_object() };
+
+        if (type_object.tp_flags & ffi::Py_TPFLAGS_READY) == 0 {
+            // automatically initialize the class on-demand
+            let gil = Python::acquire_gil();
+            let py = gil.python();
+
+            initialize_type::<Self>(py, <Self as PyTypeInfo>::MODULE).unwrap_or_else(|e| {
+                e.print(py);
+                panic!("An error occurred while initializing class {}", Self::NAME)
+            });
+        }
+
+        unsafe { NonNull::new_unchecked(type_object) }
+    }
+}
+
+/// `PyClassShell` represents the concrete layout of our `#[pyclass]` in the Python heap.
+#[repr(C)]
+pub struct PyClassShell<T: PyClass> {
+    ob_base: <T::BaseType as PyTypeInfo>::ConcreteLayout,
+    pyclass: ManuallyDrop<T>,
```

Comment:
This `PyClassShell` idea is nice, and is heading in the direction that `pybind11` goes. There's two main differences between this idea and what `pybind11` does:

1. The `pybind11` equivalent just stores `*mut void` instead of being generic over `T`. This means that the python type allocation for all pybind11 classes is more or less exactly the same. It also means smaller binary sizes, because there's only one copy of the `PyClassShell` methods instead of one for each `T`.
1. `pybind11` allows references to be stored in the `PyClassShell` as well as owned values. In Rust speak, instead of `pyclass: ManuallyDrop<T>`, we'd have something like:

       enum PyClassValue<T> {
           Borrowed(*const T),
           MutBorrowed(*mut T),
           Owned(ManuallyDrop<T>),
       }

The great thing about storing references in `PyClassShell` is that it allows us to do a lot less copying when going Rust -> Python -> Rust.

For example, the code below doesn't currently compile with `pyo3` because the `&Address` return type. If we could store references in the `PyClassShell`, then this should be fairly straightforward to support:

    #[pyclass]
    struct Address {
        value: String
    }

    #[pyclass]
    struct Employee {
        address: Address
    }

    #[pymethods]
    impl Employee {
        pub fn address(&self) -> &Address {
            &self.address
       }
    }